stan/math/prim/meta/holder.hpp

```diff
@@ -0,0 +1,319 @@
+#ifndef STAN_MATH_PRIM_FUN_HOLDER_HPP
+#define STAN_MATH_PRIM_FUN_HOLDER_HPP
+
+#include <stan/math/prim/meta/is_eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+/**
+ * \defgroup eigen_expressions Eigen expressions
+ */
+/**
+ * \ingroup eigen_expressions
+ * \defgroup returning_expressions Returning expressions
+ */
+/**
+ * \ingroup returning_expressions
+ *
+ * Operations in Eigen Expressions hold their arguments either by value or by
+ * reference. Which one is chosen depends on type of the argument. Other
+ * operations are held by value. "Heavy" objects that can hold data themselves,
+ * such as `Eigen::Matrix` or `Eigen::Ref` are instead held by reference. THis
+ * is the only criterion - holding rvalue arguments by value is not supported,
+ * so we can not use perfect forwarding.
+ *
+ * When returning an expression from function we have to be careful that any
+ * arguments in this expression that are held by reference do not go out of
+ * scope. So a function returning an expression referencing local matrices or
+ * matrices that were rvalue reference arguments to the function will not work.
+ *
+ * A workarount to this issue is allocating and constructing or moving such
+ * objects to heap. `Holder` object is a no-op operation that can also take
+ * pointers to such objects and release them when it goes out of scope. It can
+ * be created either by directly supplying pointers to such objects to `holder`
+ * function or by forwarding function arguments and moving local variables to
+ * `make_holder`, which will move any rvalues to heap first.
+ */
```

Comment:
I took this and included an example so we could describe what causes the failure in the example and how you would use `make_holder` to fix the code. What causes the need for holder seems like a weird issue most people would not have seen before so I want to make sure new developers can see this explainer and get what's happening. How do you feel about the below? Not sure if I got the explanation right but feel free to modify whatever

https://gist.github.com/SteveBronder/84c710c3f355eb53aa946983e48289a9